単体テスト、結合テストの実装

ShotaTsuru opened the pull request:
# What
単体テスト、結合テストの実装
# Why
レビューをいただくため

ShotaTsuru committed:
単体、結合テストの実装

betachelsea commented on Gemfile:
```diff
@@ -64,4 +64,6 @@ gem 'devise'
 gem 'kaminari'
 
+gem 'factory_bot'
+gem 'factory_bot_rails'
```
`factory_bot` と `factory_bot_rails` は本番環境でつかうgemではないので、groupの中に入れるようにしましょう。

引用: [thoughtbot/factory\_bot\_rails: Factory Bot ♥ Rails - Configuration](https://github.com/thoughtbot/factory_bot_rails#configuration)

```ruby
group :development, :test do
  gem 'factory_bot_rails'
end
```

また、`factory_bot_rails` のなかに `factory_bot` も含む（依存している）ので、`factory_bot` は書かなくても大丈夫です。 書くのもNGというわけではないので補足まで！ :pray:

ShotaTsuru replied:
承知致しました！

betachelsea commented on spec/support/factory_bot.rb:
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+RSpec.configure do |config|
```
今回使用しているのはminitestなので、RSpec(別のテストフレームワーク)の設定は不要かな？と思いました〜 👀

ShotaTsuru replied:
ですよね😅😅😅
使わないかも？念の為？？みたいな感じで入れてしまいました。

betachelsea commented on test/models/report_test.rb:
```diff
@@ -3,7 +3,15 @@
+  test 'Report#editable?' do
+    user = User.create(email: 'foo@example.com', name: '', password: '111111', password_confirmation: '111111')
+    report = user.reports.create(title: 'test', content: 'aaaaaaaa')
+    assert_equal true, report.editable?(user)
```
`editable?` が false を返すことも検証されていてほしいです〜 👀 
また、FactoryBotが定義されていますので、それを使うとpassword周りの設定は省略して user を作ることはできそうだと思いました！

ShotaTsuru replied:
gemを:development, :testに入れておらず使えていないことに気づいておらず、放置しておりました😭

betachelsea commented on test/models/user_test.rb:
```diff
@@ -3,7 +3,41 @@
-  # test "the truth" do
-  #   assert true
-  # end
+  include FactoryBot::Syntax::Methods
```
このテストでは FactoryBot を使っていないようです 👀 
せっかくなので、`User.create` の部分で利用すると良いと思いました！

```ruby
user = User.create(email: 'foo@example.com', name: '', password: '111111', password_confirmation: '111111')

# ↓

user = FactoryBot.create(:user, email: 'foo@example.com')
```


また、この「include FactoryBot::Syntax::Methods」が必要になるシーンは、`FactoryBot` の呼び出しを省略してかきたい場合のみです。


```ruby
FactoryBot.create(:user)
```

↓ 次のように省略できるようになる

```ruby
include FactoryBot::Syntax::Methods
create(:user) # "FactoryBot" と先頭にいちいちつけなくて良くて便利！
```

もしテスト全体で `create(:user)` という表記を活用したい場合は、`test/test_helper.rb`で呼び出すと便利です！
[FactoryBot\(FactoryGirl\)チートシート \- Qiita](https://qiita.com/morrr/items/f1d3ac46b029ccddd017)

ShotaTsuru replied:
設定したのに使えないなーと思っていたのですが、gemを:development, :testで入れてなかったので使えておリませんでした😭😭

初歩的な間違いが引き起こした放置ミスです失礼いたしました！

betachelsea commented on test/models/user_test.rb:
```diff
@@ -3,7 +3,41 @@
+    user =  User.create(email: 'foo@example.com', name: '', password: '111111', password_confirmation: '111111')
+    user2 = User.create(email: 'boo@example.com', name: '', password: '111111', password_confirmation: '111111')
+    user2.active_relationships.find_or_create_by!(following_id: user.id)
+    assert_equal true, user2.following?(user)
```
false の検証もしたいです！

ShotaTsuru replied:
承知しました！

betachelsea commented on test/models/user_test.rb:
```diff
@@ -3,7 +3,41 @@
+    user =  User.create(email: 'foo@example.com', name: '', password: '111111', password_confirmation: '111111')
+    user2 = User.create(email: 'boo@example.com', name: '', password: '111111', password_confirmation: '111111')
+    user2.passive_relationships.find_or_create_by!(follower_id: user.id)
+    assert_equal true, user2.followed_by?(user)
```
false の検証もしたいです！

ShotaTsuru replied:
承知しました！

betachelsea commented on test/models/user_test.rb:
```diff
@@ -3,7 +3,41 @@
+  test 'User#follow' do
+    user =  User.create(email: 'foo@example.com', name: '', password: '111111', password_confirmation: '111111')
+    user2 = User.create(email: 'boo@example.com', name: '', password: '111111', password_confirmation: '111111')
+    assert_equal user.active_relationships.find_or_create_by!(following_id: user2.id), user.follow(user2)
```
`user.active_relationships.find_or_create_by!(following_id: user2.id)` は `user.follow(user2)` の実行結果が実行しようとしていることそのままなので、もし `user.follow(user2)` が例えば「フォローの状態を確認してくるだけで、実はフォロー作成の機能をつけ忘れていた 」ような状態でもpassしてしまうと思います。

こういった、「状態を変化させる」ようなメソッドの検証をするときは、`user.follow(user2)` の実行結果自体を確認するのではなく、「実行前後でどう変化したか」を確認してみましょう〜 👀  ( Unfollowのテストはうまくできていますね！ 👍  )

ShotaTsuru replied:
了解しました！

betachelsea commented on test/system/reports_test.rb:
```diff
@@ -0,0 +1,51 @@
+  end
+
+  test 'visiting the Report' do
+    click_on '日報'
+    visit reports_path
```
`click_on '日報'` と `visit reports_path` は、手段は違うものの同一ページへの遷移を示しているので、どちらかのみの操作でOKそうです。

ShotaTsuru replied:
了解しました！

betachelsea commented on test/system/reports_test.rb:
```diff
@@ -0,0 +1,51 @@
+    fill_in 'report[content]', with: 'test test content'
+    click_on '登録する'
+
+    assert_text '日報が作成されました。'
```
意図通りのタイトル、コンテンツが登録されたか（表示されているか）も確認するとより確実です！
現在は「作成されました」という表示は作ったけど実は登録していなかった... というバグが見逃されうる状態かなと思います〜 👀 
他の項目も同じ観点から見直しをお願いします！ 🙏

ShotaTsuru replied:
承知しました！

betachelsea reviewed (commented):
ご確認よろしくお願いします！ 🙏

ShotaTsuru committed:
gemを適切な記述行へ変更
モデル単体テスト修正
・メソッドに対してfalseの検証を追加
・followメソッドの検証修正
結合テストの修正
・各処理後に正しくビューへ反映されているかの検証を追加

ShotaTsuru committed:
rubocop修正

betachelsea commented on test/models/report_test.rb:
```diff
@@ -3,7 +3,17 @@
+    user = User.create(email: 'foo@example.com', name: '', password: '111111', password_confirmation: '111111')
+    user2 =  FactoryBot.build(:user)
+    report = user.reports.create(title: 'test', content: 'aaaaaaaa')
+    assert_not report.editable?(user2)
```
[余談]検証はこれで大丈夫そうです！ 👍 
次の点がちょっと気になったので共有します〜 👀 

- `user` がレコード保存されている状態(create) に対して、`user2` が保存されていない(bulid、idが付与されていない=DBに記録されていない)状態なので、その差異がテスト結果に影響しないだろうか。※今回のケースでは影響しないので、事実上問題ないが、ぱっと見で少し気になる状態となっている。
- `assert_not` との対称は `assert` かなと思うが、もう片方が `assert_equal` となっていて少し違和感がある。どちらも `assert_equal` で統一されるか、`assert_not` & `assert` のセットで検証するか、が良いかなと感じる。

テスト内容ではなく書き方に関する重箱の隅の話ですが、厳しめのレビュアーさんにチェックしてもらうと指摘されそうだなと思いまして、念の為共有まで！ 🙏

ShotaTsuru replied:
おっしゃる通りです！！createとbuildは統一すべき点です！🤩直ちに修正します！
assert_notとassertの対の関係は全く知識になかったので頭に入れて修正します！

betachelsea commented on test/models/user_test.rb:
```diff
@@ -3,7 +3,44 @@
-  #   assert true
-  # end
+  test 'User#name_or_email' do
+    user = FactoryBot.create(:user, name: '')
+    assert_equal '1111@example.com', user.name_or_email
```
[余談]emailの検証をしているので、初期値がすでにそうなっていたとしても、`FactoryBot.create(:user, name: '', email: '1111@example.com')` としておくと便利です。もし将来何らかの理由でFactoryBotの内容を修正しないといけない場合があったとしても、このテストが落ちるのを防ぐことができます 👍

ShotaTsuru replied:
確かにです！

betachelsea commented on test/models/user_test.rb:
```diff
@@ -3,7 +3,44 @@
+  test 'User#follow' do
+    user =  FactoryBot.create(:user, email: 'foo@example.com')
+    user2 = FactoryBot.create(:user, email: 'boo@example.com')
+    user.follow(user2)
+    assert_equal user2.followers[0], user
```
[修正依頼] こうして修正していただいて改めて見ると、やはり「 `follow` 実行前の user2.followers[0] に元々userが入っていたということは考えられないか？」と実行前の状況が気になってしまいました 👀   実行前の検査もお願いできますでしょうか！ 🙏 

※unfollowのテストでは、 `user.active_relationships.find_or_create_by!(following_id: user2.id)` が実行されることで「必ずデータがある」状態→ `user.unfollow(user2)` 実行 →データが無くなった、ということが把握できるようになっていると考えました。

ShotaTsuru replied:
承知しました！

ただ、
実行前で`assert_equal nil, user2.followers[0]` をつけると
`user.follow(user2)`の後の
`user2.followers[0]`の結果がなぜか空のの配列が返ってきたため、followed_by?メソッドで検証いたしました。

betachelsea replied:
> user2.followers[0]の結果がなぜか空のの配列が返ってきた

なるほど。それは user2 のインスタンスが古い情報を保持したままだったからかなと思われます。
もし user2.follwers[0] を確認したい場合は、 `user2.reload` をすることで確認できたかなと考えます。
もちろん、followed_by? の変化を確認することでも問題ないのでOKです！ 👍

betachelsea commented on test/factories/users.rb:
```diff
@@ -0,0 +1,13 @@
+    postal_code            { '111-1111' }
+    address                { 'test' }
+    self_introduction      { 'testtest' }
+    email                  { '1111@example.com' }
```
[余談]こちら、 `FactoryBot.create(:user)` をしようとすると多分emailの重複でエラーとなるため、適宜 create するときは emailを指定していただいているのかな...と思いました 👀 
便利な方法として、sequence というものが使えますので共有します！
https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#inline-sequences

```ruby
sequence(:email) { |n| "person#{n}@example.com" }
```

と書いておくと、

```ruby
user =  FactoryBot.create(:user) #=> user.email は "person1@example.com"
user2 = FactoryBot.create(:user) #=> user2.email は "person2@example.com"
```

と自動で n に数字を入れてユニークにしてくれます。

ShotaTsuru replied:
なるほど！実装致します！

betachelsea commented on test/system/reports_test.rb:
```diff
@@ -0,0 +1,55 @@
+
+    assert_text '日報が作成されました。'
+    click_on '戻る'
+    page.has_text? 'test title'
```
[修正依頼] `page.has_text? 'test title'` は true/false を返しますが、「検査していない」ので、
`page.assert_text 'test title'` を使いましょう〜 👀 
他の似たような箇所も同様に修正をお願いします 🙏 

※実際に保存されているデータのタイトルは `test titele` になってしまっていました。合わせてtypo修正してもらうと良さそうです！

ShotaTsuru replied:
検査していないので、実行してもテストが通っていたように見えてテストが落ちてなかったのですね！！申し訳ありませんテストの機能把握不足でした。。

betachelsea reviewed (commented):
気になった点(余談)と修正してほしい点についてコメントしました！

また、余談ですが、commit f033ad66 に詳細なコメントをつけていただいているようですね！
コメントをみると、複数の観点の変更が1commitにまとまっているように感じました。適宜わけていくようにするのもありかもしれないです。あくまでも個人の好みやプロジェクト単位の方針によるので、余談まで！

参考: [【初心者向け】「コミットの粒度がわからない問題」の模範解答を考えてみた \- Qiita](https://qiita.com/jnchito/items/40e0c7d32fde352607be)

ShotaTsuru committed:
reportモデルのテスト修正
インスタンスの作成をcreateに変更
assert_equalをassert_notに合わせるためにassertのみに変更

ShotaTsuru committed:
コードの簡略化とスペルミスの修正、検査が適切でないテストの修正。

betachelsea reviewed (approved):
修正ありがとうございました！確認OKです〜 🎉
